Android Support

zibs opened the pull request:
This PR adds support for using Fingerprint Authentication on Android. 

It keeps the interface mostly the same, aside from allowing the user to pass in a config object to change the title and color of the pop-up requesting authentication. 

The error handling is not as robust as iOS, but future PRs would be welcome to improve this.

Readme will be updated with an image showing it working on Android shortly.

I've tested this by creating a new project with `react-native init` for `"react-native": "0.48.1",` and it seems to be working for both iOS and Android.

thanks to @sunto for doing the bulk of the work!

### Test Plan: 
- [x] - iOS still works as expected. (no changes were made to any iOS code, so this should be fine)
- [x] - Android auth works when `isSupported` is true
- [x] - Calling authenticate when `isSupported` is false doesn't die.
- [x] - Calling `isSupported` when Android has no fingerprint reader works
- [x] - If auth is supported but no fingerprints are enrolled on phone, Android doesn't die.
- [x] - If not config object is passed into Android, it defaults to the default config. 
- [x] - If config object is passed into Android, it uses the default config. 
- [x] - Android auth works 🎉

naoufal commented:
👏This is awesome — Great work guys!

🌴On vacation right now, but will try and get to this within a week. Otherwise, I'll see if I can get someone what to help with the review.

zibs commented:
Sounds good, no rush 👍  - enjoy vacay - I'll do some additional testing this weekend and Sunny will too.

ferrannp commented on README.md:
````diff
@@ -30,16 +34,39 @@ Due to the rapid changes being made in the React Native ecosystem, we are not of
+
+On android you will need to add
+```xml
+<uses-permission  android:name="android.permission.USE_FINGERPRINT" />
````
I think you do not need it. You already add this permission to the manifest library so it will get merged with your manifest (app).

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintAuthModule.java:
```diff
@@ -0,0 +1,200 @@
+          return false;
+      }
+
+      if (ActivityCompat.checkSelfPermission(getReactApplicationContext(),
```
You do not need this block, see [USE_FINGERPRINT](https://developer.android.com/reference/android/Manifest.permission.html#USE_FINGERPRINT). This is a ` normal` permission so user cannot disable it.

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintDialog.java:
```diff
@@ -0,0 +1,142 @@
+        super.onCreate(savedInstanceState);
+
+        // Do not create a new Fragment when the Activity is re-created such as orientation changes.
+        setRetainInstance(true);
```
This is considered a bad practice. You should only retain a fragment for a dialog that returns no view, so `onCreateView` should return null. Anyway you already use a DialogFragment (they don't leak on rotation), and RN in manifest also says that do not destroy the activity so you do not need this line.

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintDialog.java:
```diff
@@ -0,0 +1,142 @@
+        setRetainInstance(true);
+        setStyle(DialogFragment.STYLE_NORMAL, android.R.style.Theme_Material_Light_Dialog);
+
+        // dialogCallback = null;
```
Forgotten comment?

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintDialog.java:
```diff
@@ -0,0 +1,142 @@
+
+        // Do not create a new Fragment when the Activity is re-created such as orientation changes.
+        setRetainInstance(true);
+        setStyle(DialogFragment.STYLE_NORMAL, android.R.style.Theme_Material_Light_Dialog);
```
Do we need to set style here explicitly? Can we get the default style? Maybe be able to override it with custom `style`?

sunto replied:
This makes sense, we should probably do something like the following in the TouchID.android.js
DEFAULT_CONFIG = { title: 'Authentication Required', color: '#1306ff', style: 'STYLE_NORMAL' };

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintDialog.java:
```diff
@@ -0,0 +1,142 @@
+            @Override
+            public void onClick(View view) {
+                mFingerprintHandler.endAuth();
+                // dismiss();
```
Forgotten comment?

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintHandler.java:
```diff
@@ -0,0 +1,83 @@
+    private final Callback mCallback;
+
+    public FingerprintHandler(Context context, FingerprintManager fingerprintManager, Callback callback) {
+        appContext = context;
```
It keep names consistent, so if we use `mCallback` shouldn't we also use `mAppContext` ?

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintHandler.java:
```diff
@@ -0,0 +1,83 @@
+                && mFingerprintManager.hasEnrolledFingerprints()
+                && (ActivityCompat.checkSelfPermission(appContext,
+                        Manifest.permission.USE_FINGERPRINT) ==
+                        PackageManager.PERMISSION_GRANTED);
```
Same permission comment as before, also, this function seems a duplicated from `FingerprintAuthModule.java` ?

ferrannp commented on android/src/main/java/com/rnfingerprint/FingerprintHandler.java:
```diff
@@ -0,0 +1,83 @@
+
+        if (ActivityCompat.checkSelfPermission(appContext,
+                Manifest.permission.USE_FINGERPRINT) !=
+                PackageManager.PERMISSION_GRANTED) {
```
Not needed.

See: https://developer.android.com/training/permissions/requesting.html

> Normal permissions do not directly risk the user's privacy. If your app lists a normal permission in its manifest, the system grants the permission automatically.

ferrannp commented on android/src/main/res/layout/fingerprint_dialog.xml:
```diff
@@ -0,0 +1,62 @@
+                  android:layout_width="49dp"
+                  android:layout_height="49dp"
+                  android:layout_marginLeft="25dp"
+                  android:src="@drawable/ic_fp_40px"
```
I would not generate a png for this icon but you can generate the xml with Android Studio. Then will look good anywhere.

ferrannp commented on android/src/main/res/layout/fingerprint_dialog.xml:
```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:orientation="vertical" android:layout_width="match_parent"
```
This file seems it just needs to run 'Reformat' through Android Studio.

ferrannp commented on android/build.gradle:
```diff
@@ -0,0 +1,34 @@
+
+  defaultConfig {
+    minSdkVersion 16
+    targetSdkVersion 22
```
I would bump `minSdkVersion` and `targetSdkVersion` up. FingerPrint works only on Android 23 and above, right? So it does not make really sense we say `16`.

ferrannp commented:
@zibs could you update the `index.android.js` into the examples folder? I actually try it and it works so good so far :). Just added some comments into the PR.

@naoufal maybe you want to change the lib name ! And support face id for iOS in the future ;) Just saying.

wireforce commented:
Maybe rename to react-native-biometric-auth ? :)

zibs commented:
@ferrannp Thanks so much for the review! Will take a look at get these updates to you shortly.

naoufal commented:
@zibs I dig it, picked up https://www.npmjs.com/package/react-native-biometric-auth.  Will add you and @sunto to the package.

GantMan commented:
Excited to see this land!

zibs commented:
landing shortly...thanks for the nudge @GantMan 🥇

harrisrobin commented:
@zibs we're super excited to see this land! Great work guys :)

tactico commented:
@zibs Also looking forward to this!

ben-snaize commented:
Is it in yet? Hyped for this!

zibs commented:
Sorry for delay all - just got one or two last things to update here and we should be goodd!

tactico commented:
I could not wait and implemented your branch and started testing. One comment would be to synchronize the error codes with the iOS version, for example in the IOS version I get an error of the form "LAErrorUserCancel" and the same error in this branch comes through as "Touch ID Error"

zibs commented:
@tactico Feel free to make a PR! The errors are handled fundamentally different on iOS and Android, which is why I haven't gone down that path yet. Definitely could be somewhat integrated eventually though!

naoufal commented:
Agreed, let's tack on error handling in a subsequent PR, since even the
current iOS error handling could use some work.

@tactico, let me know if you're up for taking this on.

On Fri, Nov 3, 2017, 10:40 AM Eli Zibin <notifications@github.com> wrote:

> @tactico <https://github.com/tactico> Feel free to make a PR! The errors
> are handled fundamentally different on iOS and Android, which is why I
> haven't gone down that path yet. Definitely could be somewhat integrated
> eventually though!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/naoufal/react-native-touch-id/pull/53#issuecomment-341776047>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABjWsLU51DqtuKBlwATSRQUAbfPY0jiWks5sy1AGgaJpZM4PdG6w>
> .
>

tactico commented:
I probably will not be able to get to it for about a week, major release of our product scheduled for this Friday!

tactico commented:
Sorry noufal, got pulled into another project so likely will not be able to do this anytime soon. My apologies...

zibs commented:
@naoufal Hey Naofal, sorry for dropping the ball and taking so long on this one. I've just pushed a number of commits that addressed most of the issues in the review. There is one missing about passing in a custom style theme thing, but I think that can wait until later!

I just went through and tested on a brand new react native project created with `react-native init` - I wasn't able to run the example cause the `yarn.lock` is pointing to netflix artifacts 🎥 . But I added the example code I used in my test run and it should be all good!

Let me know if you have any suggestions!

naoufal commented:
No worries at all!  This is great work that a bunch of people will benefit from, I'd say you're doing the total opposite of dropping the ball! 🍻 

Sorry about the netflix artifacts -- I'll fix that!

I'll take a look tonight/tomorrow evening and try and get us setup for a merge this weekend.

naoufal commented:
@zibs updated the `yarn.lock`.  Also merged the FaceID PR so you'll need to update the README.  Once that's done hit the merge button!

And I'll bake this into the next release 🎉

fretman92 commented on TouchID.android.js:
```diff
@@ -1,13 +1,53 @@
+  },
+
+  authenticate(reason, config) {
+    DEFAULT_CONFIG = { title: 'Authentication Required', color: '#1306ff' };
```
why do you not declare the `DEFAULT_CONFIG` variable here?
It causes error in strict mode ?
